### PR TITLE
Add pod support

### PR DESCRIPTION
**Pull request by dfturn**
New in Marathon 1.4.0: https://github.com/mesosphere/marathon/blob/master/changelog.md#changes-from-1310-to-140

This adds support to query pods. Testing done on a Marathon deployment with version 1.4.2. 

Addresses issue https://github.com/gambol99/go-marathon/issues/271

### CODE REVIEW

**Comment by dfturn** [comment]
Hmm that TestMarkDown appears to have a race condition as the `cluster.markDown` function starts a goroutine.

**Comment by timoreimann** [comment]
Pod support in go-marathon, fantastic! :tada:

This PR is fairly big, so I'll need some time to work through everything.

On a first glance, it looks like traditional applications and pods share a number of APIs. Can we consider extracting a common interface to possibly share or at least minimize the amount of (now duplicated) code?

**Comment by timoreimann** [comment]
The test failure is likely a flake that #296 should address. I'll try to kick Travis again.

**Comment by dfturn** [comment]
Thanks for looking @timoreimann! 

Because of the [pod and application spec differences](https://github.com/mesosphere/marathon/tree/master/docs/docs/rest-api/public/api/v2/types), most of the code which appears to be duplicated is difficult to de-dup. I'm happy to take recommendations though! 

In terms of interfaces, we could create some sort of `Service` interface that encompasses Pods and Applications; you'd retrieve one with the current marathon client but could then perform actions using that object itself, e.g. `s := client.GetPod("foo"); s.Scale(5)`. I like the idea of performing actions on a Service object as opposed to a single mega client, but otherwise the interface itself feels a little low value to me, particularly as I don't see more "Service" types being created in Marathon.

**Comment by timoreimann** [comment]
@dfturn in general, I'd be okay with having only two implementations of the Service for the foreseeable future if it helps us to cut down on the amount of code we'd have to manage. I'd probably have to reserve some time to better see if clever usage of interfaces could help with  significantly dropping the code volume this PR brings -- maybe you already have a better feel for how much of a gain that would be?

There are also other alternatives we can consider; like generating, or loosening type safety and go with more `interface{}`s. They are definitely not my first-picks, however. I'd just like us to consider all reasonable options prior to committing to adding 2k LOC for pod support.

Maybe @gambol99 and/or @marco-jantke may want to chime in with ideas on how to handle the new code best.

Thanks again for making the motion to bringing in pod support. 👏  Looking forward to the merge already. 😃

**Comment by dfturn** [comment]
@timoreimann @gambol99 @marco-jantke 

Hi guys, just pinging for some feedback on this PR. I know it's pretty large but I'd love to get this in!

**Comment by timoreimann** [comment]
@dfturn I'd like to explore the idea of introducing an interface (as we discussed before) to reduce the amount of duplicated code. Do you think you could give it enough of a try just to see if the approach pays off?

**Comment by dfturn** [comment]
I did a little exploration in commit a26c160 but didn't really like how it felt, and I don't think it will reduce any of the duplication for the reasons I described before. To make it work we have to sneakily embed the marathon client in the `Service` instance in `Get` and `Create` calls, and only then can we do direct calls with the `Service` object with the hidden client. But it's tricky because it won't be transparent to the user, since if they tried to create a new pod in code then call `Create` on it, it wouldn't work without the embedded client being initialized. So the workflow is odd; some calls you have to do with the client and some with the `Service` object.

```
client := marathon.NewClient()
pod := client.GetPod("pod_id")
pod.Scale(5)

podNew := NewPod().Name("second_pod_id") // do some other things to actually create it
client.CreatePod(podNew)
```

**Comment by marco-jantke** [comment]
Sorry for the very late response.. 

I agree with @dfturn that the interface would be rather awkward to use through the introduction of a `Service` interface. I also checked whether we could internally leverage some code reuse, but this seems hard as well through the API differences mentioned above. Have a look at the following examples:

```
func (r *marathonClient) GetPodByVersion(name, version string) (*Pod, error) {
	uri := fmt.Sprintf("%s/%s", buildPodVersionURI(name), version)
	result := new(Pod)
	if err := r.apiGet(uri, nil, result); err != nil {
		return nil, err
	}

	return result, nil
}
....
func (r *marathonClient) ApplicationByVersion(name, version string) (*Application, error) {
	app := new(Application)

	path := fmt.Sprintf("%s/versions/%s", buildPath(name), version)
	if err := r.apiGet(path, nil, app); err != nil {
		return nil, err
	}

	return app, nil
}
```
AFAICS there is no abstraction that we could derive from these samples without losing their context. I could only think about a signature like this: `UnitByVersion(path string, result interface{}) (interface{}, error)` which would make it even more difficult for the clients.

Long story short, I don't see a useful to reuse a lot of code.

**Comment by dfturn** [comment]
Appreciate you chiming in @marco-jantke. Could I get some eyes on the specifics of the PR so we can move toward merging it?

**Comment by timoreimann** [comment]
@dfturn sorry for the awfully long delay, been running short on bandwidth lately.

I'll make sure this PR gets back on my radar and try to review it some time next week.

**Comment by dfturn** [comment]
Hi @timoreimann, any chance we can get a review going on this PR?

**Comment by timoreimann** [comment]
hey there @dfturn. Sorry I couldn't keep up with my promise to review soon. :( I have actually already started the review but I'm yet to finish. Hopefully I can get this done soon.

File: client.go [hunk 1]
```diff
@@ -299,9 +340,23 @@ func (r *marathonClient) apiCall(method, path string, body, result interface{})
...
-					return fmt.Errorf("failed to unmarshal response from Marathon: %s", err)
+				// If we have a deployment ID header and no response body, give them that
+				deploymentID := response.Header.Get(deploymentHeader)
+				if len(respBody) == 0 && deploymentID != "" {
+					d := DeploymentID{
+						DeploymentID: deploymentID,
+					}
+					switch result.(type) {
+					case *DeploymentID:
+						*result.(*DeploymentID) = d
```
Comment (timoreimann) [inline client.go]: We could postpone the definition of `d` until we're clear that `result` is of the right type, or even just assign the `DeploymentID` struct inline here.

**Comment by dfturn** [comment]
I appreciate all of the feedback! I'll start going through it. 

For your overall comments:
- omitempty: I'll make a thorough pass. 
- convenience methods: when I originally created this fork, I was using the pod structures without any convenience methods. I can definitely say it feels a lot cleaner to interact with the library while using them. But as an owner on the repo, if you have preferences there, I can trim them down.

**Comment by timoreimann** [comment]
Prior to removing the convenience methods, could you please first add a Pod usage example to the examples folder? That might give a clearer picture of their value.

**Comment by timoreimann** [comment]
Sorry for the lack of attention. Github actually doesn't always send out notifications when commits are just pushed. It's usually helpful to me to leave an extra comment, also to know that a PR is ready for review again.

I suppose I can have another look, right?

**Comment by dfturn** [comment]
Hey @timoreimann, apologies, that commit is actually unrelated to this PR. I've gotten distracted myself and need to give this PR some attention. I'll address your feedback this weekend and let you know when it's ready for another look.

**Comment by timoreimann** [comment]
@dfturn no worries, appreciate the heads up.

Looking forward to wrapping things up soon.

File: network.go [hunk 2]
```diff
@@ -0,0 +1,103 @@
...
+	return &PodEndpoint{
+		Protocol: []string{},
+		Labels:   map[string]string{},
+	}
+}
+
+// NewContainerPodNetwork creates a PodNetwork for a container
+func NewContainerPodNetwork(name string) *PodNetwork {
+	pn := NewPodNetwork(name)
+	return pn.SetMode("container")
```
[the remark this answers is not in the document]
Comment (dfturn) [reply network.go]: This is just for convenience of creating a container network, which is at least my primary use case. `NewPodNetwork` is still available for other modes.

File: pod.go [hunk 3]
```diff
@@ -0,0 +1,310 @@
...
+		Networks:    []*PodNetwork{},
+	}
+}
+
+// String marshals the pod as an indented string
+func (p *Pod) String() string {
+	s, err := json.MarshalIndent(p, "", "  ")
+	if err != nil {
+		return fmt.Sprintf(`{"error": "error decoding type into json: %s"}`, err)
+	}
```
[the remark this answers is not in the document]
Comment (dfturn) [reply pod.go]: I've tended to use it for debugging, yes. This does match the implementation of the Application.String() method: https://github.com/gambol99/go-marathon/blob/master/application.go#L625

I also tend to like it as JSON because that's how I normally interact with these objects on the filesystem, and then I can dump them in the UI for more rapid testing.

File: pod_container.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+package marathon
+
+// PodContainer describes a container in a pod
+type PodContainer struct {
+	Name         string                 `json:"name,omitempty"`
+	Exec         *PodExec               `json:"exec,omitempty"`
+	Resources    *Resources             `json:"resources,omitempty"`
+	Endpoints    []*PodEndpoint         `json:"endpoints,omitempty"`
+	Image        *PodContainerImage     `json:"image,omitempty"`
+	Environment  map[string]interface{} `json:"environment,omitempty"`
```
[the remark this answers is not in the document]
Comment (dfturn) [reply pod_container.go]: Yes. It's simply for convenience, some environment variables may be applicable to all containers.

File: client.go [hunk 5]
```diff
@@ -299,9 +340,23 @@ func (r *marathonClient) apiCall(method, path string, body, result interface{})
...
+					}
```
[the remark this answers is not in the document]
Comment (dfturn) [reply client.go]: Thanks for the cleanup suggestion, converting to that.

File: client.go [hunk 6]
```diff
@@ -299,9 +340,23 @@ func (r *marathonClient) apiCall(method, path string, body, result interface{})
 
 		// step: check for a successfull response
 		if response.StatusCode >= 200 && response.StatusCode <= 299 {
+
 			if result != nil {
-				if err := json.Unmarshal(respBody, result); err != nil {
-					return fmt.Errorf("failed to unmarshal response from Marathon: %s", err)
+				// If we have a deployment ID header and no response body, give them that
+				deploymentID := response.Header.Get(deploymentHeader)
+				if len(respBody) == 0 && deploymentID != "" {
```
[the remark this answers is not in the document]
Comment (dfturn) [reply client.go]: That's true for PUT, but DELETE only returns the deployment ID in the header. As far as I can tell, that's the only flow in which this occurs, since otherwise the pod is contained in the body. Since in other situations we have other ways of determining the status of the deployment, I'm not too concerned about losing the deployment ID to the pod body, but I'm open to feedback there.

File: health.go [hunk 7]
```diff
@@ -30,6 +30,35 @@ type HealthCheck struct {
...
+// PodHealthCheck describes how to determine a pod's health
+type PodHealthCheck struct {
+	HTTP                   *HTTPHealthCheck    `json:"http,omitempty"`
+	TCP                    *TCPHealthCheck     `json:"tcp,omitempty"`
+	Exec                   *CommandHealthCheck `json:"exec,omitempty"`
+	GracePeriodSeconds     int                 `json:"gracePeriodSeconds,omitempty"`
+	IntervalSeconds        int                 `json:"intervalSeconds,omitempty"`
+	MaxConsecutiveFailures int                 `json:"maxConsecutiveFailures,omitempty"`
+	TimeoutSeconds         int                 `json:"timeoutSeconds,omitempty"`
+	DelaySeconds           int                 `json:"delaySeconds,omitempty"`
```
[the remark this answers is not in the document]
Comment (dfturn) [reply health.go]: I think you are correct, that these values could be potentially valid with zero. I will note that a couple of the variables in the app `HealthCheck` above are not pointerized. I'm not changing them to not break backwards compatibility.

**Comment by dfturn** [comment]
@timoreimann I've pushed another commit which should handle your feedback. I'm not sure how to fix the test failures in Travis, since it seems to be overriding my import of the `dfturn` fork in the pod example to get my changes pulled in. 

Per your request for an example of the convenience methods, I've added that into the `examples/pods` directory. As a user, using the methods feels much cleaner to me than interacting with raw structs. 

Regarding the secrets implementation in #299, I see no way to use that marshaling logic without copy/pasting and updating for pods. And I don't think that it buys the end user much. While we do have an `interface{}`, those convenience methods described above should abstract that away from the user. For instance, to add a secret to a pod, you are doing the following, and never interact with the raw interface:

```golang
pod := marathon.NewPod()
pod.AddFullSecret("secret0", "APP_ENVIRONMENT_VARIABLE", "APP_SECRET_KEY")
```

Thanks again for all of the feedback and helping me improve this PR!

**Comment by tobilg** [comment]
Any news on this? I'd like to use the pod functionality for https://github.com/virtual-kubelet/virtual-kubelet/issues/15 Thanks!

**Comment by dfturn** [comment]
@timoreimann I'd appreciate another pass on this PR when you've got a chance.

File: examples/pods/main.go [hunk 8]
```diff
@@ -0,0 +1,170 @@
...
+	assert(err)
+
+	// Scale it
+	fmt.Println("Scaling pod...")
+	pod.Count(5)
+	pod, err = client.UpdatePod(pod, true)
+	assert(err)
+
+	// Get instances
+	status, err := client.PodStatus(pod.ID)
```
Comment (timoreimann) [inline examples/pods/main.go]: Travis CI [fails](https://travis-ci.org/gambol99/go-marathon/jobs/301150639) on this line with

`./main.go:141: client.PodStatus undefined (type marathon.Marathon has no field or method PodStatus)`

File: health.go [hunk 9]
```diff
@@ -30,6 +30,35 @@ type HealthCheck struct {
...
+// PodHealthCheck describes how to determine a pod's health
+type PodHealthCheck struct {
+	HTTP                   *HTTPHealthCheck    `json:"http,omitempty"`
+	TCP                    *TCPHealthCheck     `json:"tcp,omitempty"`
+	Exec                   *CommandHealthCheck `json:"exec,omitempty"`
+	GracePeriodSeconds     *int                `json:"gracePeriodSeconds,omitempty"`
+	IntervalSeconds        *int                `json:"intervalSeconds,omitempty"`
+	MaxConsecutiveFailures *int                `json:"maxConsecutiveFailures,omitempty"`
+	TimeoutSeconds         *int                `json:"timeoutSeconds,omitempty"`
+	DelaySeconds           *int                `json:"delaySeconds,omitempty"`
```
[the remark this answers is not in the document]
Comment (dfturn) [reply health.go]: I did make a pass through these fields per your comments in the previous round of feedback. From my interpretation of the spec, these should be pointerized, For example, `gracePeriodSeconds` defaults to 300 but the minimum is zero and as such you may want to set it to that.

File: pod.go [hunk 10]
```diff
@@ -0,0 +1,310 @@
...
+	str := ""
+	var err error
+
+	if val, ok := p.Environment[name]; ok {
+		switch val.(type) {
+		case string:
+			str = val.(string)
+			err = nil
+		case EnvironmentSecret:
+			err = fmt.Errorf("environment variable refers to a secret")
```
[the remark this answers is not in the document]
Comment (dfturn) [reply pod.go]: I made this comment responding to your secret implementation recommendation: https://github.com/gambol99/go-marathon/pull/300#issuecomment-315593015

To pull it into this thread:

Regarding the secrets implementation in #299, I see no way to use that marshaling logic without copy/pasting and updating for pods. And I don't think that it buys the end user much. While we do have an interface{}, those convenience methods described above should abstract that away from the user. For instance, to add a secret to a pod, you are doing the following, and never interact with the raw interface:

```
pod := marathon.NewPod()
pod.AddFullSecret("secret0", "APP_ENVIRONMENT_VARIABLE", "APP_SECRET_KEY")
```

File: pod_instance.go [hunk 11]
```diff
@@ -0,0 +1,116 @@
...
+	HostPorts   []int       `json:"hostPorts"`
+	IPAddresses []IPAddress `json:"ipAddresses"`
+}
+
+// String marshals the pod as an indented string
+func (p *PodInstance) String() string {
+	s, err := json.MarshalIndent(p, "", "  ")
+	if err != nil {
+		return fmt.Sprintf(`{"error": "error decoding type into json: %s"}`, err)
+	}
```
[the remark this answers is not in the document]
Comment (dfturn) [reply pod_instance.go]: I don't feel strongly about it, I've just removed the `String()` implementations to have default print formatting then.

**Comment by dfturn** [comment]
@timoreimann made another pass, Replied to a couple of the threads as well, specifically about secrets and pointerized fields.

File: pod.go (hunk 10 above)
[the remark this answers is not in the document]
Comment (timoreimann) [reply pod.go]: I agree that we'd be able to hide the `interface{}` behind the convenience method. However, it still makes it harder for readers which now have to do type-casting to access the values, and do an additional check if they want to separate secrets from non-secrets environment variables. I have given a [more elaborate answer](https://github.com/gambol99/go-marathon/pull/299#discussion_r129804187) to the design question in PR #299.

Another point is that applications and pods would now handle environment variables and secrets differently, which I think is a good reason to continue following the path taken in PR #299.

File: health.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (timoreimann) [reply health.go]: Thanks 👍

**Comment by timoreimann** [comment]
The secrets and the general question on whether we should keep those convenience methods or not is all that's left at this point. And even though we haven't crossed the finishing line yet: @dfturn, great work -- thanks a ton for all the work you've invested. 👏

**Comment by timoreimann** [comment]
@gambol99 do you have time to comment on the question of whether we should introduce convenience methods for basically all new Pod-related fields or just those that are managed by pointers? It's basically a trade-off between (fullI) convenience and code we need to maintain.

I noted down my initial concern in [this comment](https://github.com/gambol99/go-marathon/pull/300#pullrequestreview-60323443) with @dfturn's response to prefer keeping all convenience methods [living here](https://github.com/gambol99/go-marathon/pull/300#issuecomment-331603381). The difference in convenience can be observed [in the example file](https://github.com/dfturn/go-marathon/blob/pod_support/examples/pods/main.go). I can't tell for certain how many LOCs we'd be able to drop with all "unnecessary" convenience methods gone, but I guess it'd be quite a few hundreds taking both code and test and account.

**Comment by timoreimann** [comment]
One more thing: Please extend `.gitignore` by `examples/pods/pods`.

**Comment by dfturn** [comment]
@timoreimann I'm not sure where we're at with this PR. Are we blocked on feedback from @gambol99?

**Comment by gambol99** [comment]
eeekk .. i hope i'm not the blocker here :-) ... In regard to convenience methods I agree with the block, perhaps we over extended what was really need (i blame that fact i was coming from ruby! :-))  

> In general but also in particular for this PR, I wonder if we should really provide convenience methods for all new structs we add. It leads to a lot of new code being added, and we need to be somewhat careful not to bloat the code base. IMHO, we should settle on having convenience methods for pointerized types to ease setting and clearing values, but otherwise just expect users to initialize struct fields directly. WDYT?

But given this is a cool feature i'm tempted to merge and iterate

**Comment by timoreimann** [comment]
@gambol99 thanks for the feedback.

>But given this is a cool feature i'm tempted to merge and iterate

We have to be aware that it'll be much harder to remove those convenience methods later (should we decide to) than it is to leave them out now and add them later only. But I'm undecided myself, so unless there's a stronger opinion I'm okay with keeping them in.

Apart from that, the only point left is the secrets discussion. Per [my last comment](https://github.com/gambol99/go-marathon/pull/300#discussion_r156821363) on the subject, I really think we should stick the approach we have chosen before and make sure that secrets can be managed and accessed in a consistent way. With the convenience methods discussion out of the way, that's the only blocker left for me.

**Comment by dfturn** [comment]
@timoreimann I've made the requested changes to more strongly type the `Environment` fields with secret handling. I based it off the Application marshalling implementation, but there's not a huge amount of room for de-duplication as some parts change for both the `Pod` and `PodContainer` (since secrets can be stored at container level, discrete from the top-level pod). 

And I believe the test failures I'm getting are unrelated to my changes (I merged master):
```

**Comment by timoreimann** [comment]
The [last master build was green](https://travis-ci.org/gambol99/go-marathon/builds/320612162), so I guess something is (semantically) conflicting between this PR and master...

**Comment by timoreimann** [comment]
I believe this is an issue with the latest stretchr/testify version: It worked on my local machine where I still had an older version of the testing library installed. After an update, I see the error as well.

Looks like we finally need to start pinning dependencies for the CI run.

**Comment by timoreimann** [comment]
Indeed, the library's behavior changed at commit stretchr/testify@88a414d carrying the title _generalize Empty assertion_.

**Comment by timoreimann** [comment]
@dfturn I pushed a commit to adjust to the new `NotEmpty()` behavior.

The build continues but [now fails](https://travis-ci.org/gambol99/go-marathon/jobs/330151548#L787-L788) due to a build problem in the examples:

```
# github.com/gambol99/go-marathon/examples/pods
./main.go:71:15: cannot use map[string]interface {} literal (type map[string]interface {}) as type map[string]string in field value
```

I'll let you deal with this one.

**Comment by dfturn** [comment]
Ah, sorry I forgot `make test` didn't run the examples. Thanks for fixing the other test failure!

Any thoughts on the latest changes or are we good to go?

**Comment by timoreimann** [comment]
I'll do a final pass over the PR and give it a quick test run in our infrastructure. The latter should be a good indicator on whether we have properly maintained backwards compatibility.

We're almost there...

File: pod.go [hunk 12]
```diff
@@ -0,0 +1,278 @@
...
+
+// Pod is the definition for an pod in marathon
+type Pod struct {
+	ID      string            `json:"id,omitempty"`
+	Labels  map[string]string `json:"labels,omitempty"`
+	Version string            `json:"version,omitempty"`
+	User    string            `json:"user,omitempty"`
+	// Non-secret environment variables. Actual secrets are stored in Secrets
+	// Magic happens at marshaling/unmarshaling to get them into the correct schema
+	Environment map[string]string `json:"-"`
```
Comment (timoreimann) [inline pod.go]: Can we call this `Env` to be consistent with the naming on the application?

We should then also make sure the convenience methods dealing with environment variables match those from the application.

File: pod.go [hunk 13]
```diff
@@ -0,0 +1,278 @@
...
+	Secrets     map[string]Secret `json:"-"`
+	Containers  []*PodContainer   `json:"containers,omitempty"`
+	//SecretSources     map[string]SecretSource `json:"secrets,omitempty"`
```
Comment (timoreimann) [inline pod.go]: Let's remove this comment.

File: pod.go [hunk 14]
```diff
@@ -87,49 +90,27 @@ func (p *Pod) SetLabels(labels map[string]string) *Pod {
...
-			err = nil
-		case EnvironmentSecret:
-			err = fmt.Errorf("environment variable refers to a secret")
-		default:
-			err = fmt.Errorf("environment variable refers to unknown type")
-		}
-	} else {
-		err = fmt.Errorf("environment variable not found")
+	if p.Environment == nil {
+		p = p.EmptyEnvironment()
```
Comment (timoreimann) [inline pod.go]: We don't really need to empty the environment because `AddEnvironment` already takes care of that, right?

File: pod.go [hunk 15]
```diff
@@ -0,0 +1,278 @@
...
+
+// GetSecretSource gets the source of the named secret
+func (p *Pod) GetSecretSource(name string) (string, error) {
+	if val, ok := p.Secrets[name]; ok {
+		return val.Source, nil
+	}
+	return "", fmt.Errorf("secret does not exist")
+}
+
+// AddSecret will add the secret to both the secret source and the environment
```
Comment (timoreimann) [inline pod.go]: s/will/adds/

Comment (timoreimann) [inline pod.go]: Is the part about "both the secret source and the environment" still true?

File: pod_container_marshalling.go [hunk 16]
```diff
@@ -0,0 +1,94 @@
...
+		switch envValOrSecret := genericEnvValue.(type) {
+		case string:
+			env[envName] = envValOrSecret
+		case map[string]interface{}:
+			for secret, secretStore := range envValOrSecret {
+				if secStore, ok := secretStore.(string); ok && secret == "secret" {
+					secrets[secStore] = Secret{EnvVar: envName}
+					break
+				}
+				return fmt.Errorf("unexpected secret field %v or value type %T", secret, envValOrSecret[secret])
```
Comment (timoreimann) [inline pod_container_marshalling.go]: I think `or` should say `of`.

The typo was already in the application equivalent. Could you please fix both?

**Comment by tobilg** [comment]
Thanks @dfturn @timoreimann @gambol99! Great work!
